src/ffi2/datetime.rs

```diff
@@ -0,0 +1,354 @@
+    pub data: [c_uchar; _PyDateTime_DATE_DATASIZE],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Time {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_TIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_DateTime {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_DATETIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Delta {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub days: c_int,
+    pub seconds: c_int,
+    pub microseconds: c_int,
+}
+
+// C API Capsule
+unsafe impl Sync for PyDateTime_CAPI {}
+
+lazy_static! {
+    pub static ref PyDateTimeAPI: PyDateTime_CAPI = unsafe { PyDateTime_IMPORT() };
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_IMPORT() -> PyDateTime_CAPI {
+    // PyDateTime_CAPSULE_NAME is a macro in C
+    let PyDateTime_CAPSULE_NAME = CString::new("datetime.datetime_CAPI").unwrap();
+
+    let capsule = PyCapsule_Import(PyDateTime_CAPSULE_NAME.as_ptr(), 0);
+    *(capsule as *const PyDateTime_CAPI)
+}
+
+//
+// Type Check macros
+//
+#[inline(always)]
+pub unsafe fn PyDate_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DateType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDate_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DateType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DateTimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DateTimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTime_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.TimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTime_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.TimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDelta_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DeltaType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDelta_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DeltaType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTZInfo_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.TZInfoType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTZInfo_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.TZInfoType) as c_int
+}
+
+//
+// Accessor functions
+//
+macro_rules! _access_field {
+    ($obj:expr, $type: ident, $field:tt) => {
+        (*($obj as *mut $type)).$field
+    };
+}
+
+// Accessor functions for PyDateTime_Date
+// Note: These have nonsensical names
+#[macro_export]
```

Comment:
Do we need to export this macro? Otherwise I'd remove the `#[macro_export]`

Reply:
Same goes for `PyDateTime_GET_MONTH` and PyDateTime_GET_DAY``

Reply:
These macros are part of the [public C API](https://docs.python.org/3/c-api/datetime.html), which is why I exposed them in the public API here.

Reply:
Hmm I'd say they should be not exported for now and export them as part of the ffi2/ffi3 module once the use_extern_macros is stabilized (which is expected with rust 2018)

Reply:
Another option is to refactor these as functions. I think it's probably OK to have the function version look like:

```rust
pub unsafe fn PyDatetime_GET_MONTH(o: *mut PyObject) -> u32 {
    (*o as PyDateTime_Date).data[2] as u32
}
```

The C memory is laid out specifically to allow this sort of type punning, I just felt a little weird about doing it in Rust.

Reply:
If functions are possible, than I'm definitely for functions.